pkg/kubelet/eviction/eviction_manager.go

```diff
@@ -136,6 +139,52 @@ func (m *managerImpl) IsUnderDiskPressure() bool {
 	return hasNodeCondition(m.nodeConditions, api.NodeDiskPressure)
 }
 
+func (m *managerImpl) createThresholdNotification(diskInfoProvider DiskInfoProvider, podFunc ActivePodsFunc, observations signalObservations) error {
+	// default to no-op notifier
+	m.thresholdNotifier = NewNoOpThresholdNotifier()
+
+	thresholds := m.config.Thresholds
+	if len(thresholds) == 0 {
+		return nil
+	}
+
+	for _, threshold := range thresholds {
+		// only enable memcg threshold notification if a hard memory eviction limit is set
+		if threshold.Signal != SignalMemoryAvailable || threshold.GracePeriod != time.Duration(0) {
+			continue
+		}
+
+		observed, found := observations[SignalMemoryAvailable]
+		if !found {
+			continue
+		}
+
+		quantity := getThresholdQuantity(threshold.Value, observed.capacity, resource.DecimalSI).String()
+		// TODO: remove hardcoded root memory cgroup path
+		cgpath := "/sys/fs/cgroup/memory"
+		attribute := "usage_in_bytes"
+		thresholdNotifier, err := NewMemCGThresholdNotifier(cgpath, attribute, quantity)
+		if err != nil {
+			return err
+		}
+
+		glog.Infof("registered memory notification for %s on %s at %s", cgpath, attribute, quantity)
+		m.thresholdNotifier = thresholdNotifier
+		go m.thresholdNotifier.Start()
+		go func() {
+			for {
+				// when we get a threshold notification, run synchronize to take action
+				m.thresholdNotifier.WaitForNotification()
+				glog.Infof("memory threshold notification recieved")
```

Comment:
debated on merit of throwing an event here as well... thoughts?  we don't know if the direction is moving above or below the threshold, just that it was crossed in *some* direction?